src/main/java/tech/stackable/spark/operator/abstractcontroller/AbstractCrdController.java

```diff
@@ -0,0 +1,318 @@
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+
+import org.apache.log4j.Logger;
+
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.fabric8.kubernetes.api.model.apiextensions.v1.CustomResourceDefinition;
+import io.fabric8.kubernetes.client.CustomResource;
+import io.fabric8.kubernetes.client.CustomResourceDoneable;
+import io.fabric8.kubernetes.client.CustomResourceList;
+import io.fabric8.kubernetes.client.DefaultKubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.dsl.MixedOperation;
+import io.fabric8.kubernetes.client.dsl.Resource;
+import io.fabric8.kubernetes.client.dsl.base.CustomResourceDefinitionContext;
+import io.fabric8.kubernetes.client.informers.ResourceEventHandler;
+import io.fabric8.kubernetes.client.informers.SharedIndexInformer;
+import io.fabric8.kubernetes.client.informers.SharedInformerFactory;
+import io.fabric8.kubernetes.client.informers.cache.Cache;
+import io.fabric8.kubernetes.client.informers.cache.Lister;
+
+/**
+ * Abstract crd Controller to work with customize crds. Applies given crd and orders events (add, update, delete)
+ * in an blocking queue. crdClass extending CustomResource and crdClassList extending CustomResourceList required!
+ *
+ * @param <Crd>         - pojo extending CustomResource
+ * @param <CrdList>     - pojo extending CustomResourceList
+ * @param <CrdDoneable> - pojo extending CustomResourceDoneable
+ */
+public abstract class AbstractCrdController<
+  Crd extends CustomResource,
+  CrdList extends CustomResourceList<Crd>,
+  CrdDoneable extends CustomResourceDoneable<Crd>>
+  implements Runnable {
+
+  private static final Logger LOGGER = Logger.getLogger(AbstractCrdController.class.getName());
+
+  private static final Integer WORKING_QUEUE_SIZE = 1024;
+
+  private final BlockingQueue<String> blockingQueue;
+  private final String crdPath;
+  private final Long resyncCycle;
+
+  private List<HasMetadata> crdMetadata;
+  private CustomResourceDefinitionContext crdContext;
+  private String namespace;
+
+  private KubernetesClient client;
+  private SharedInformerFactory informerFactory;
+
+  private SharedIndexInformer<Crd> crdSharedIndexInformer;
+  private Lister<Crd> crdLister;
+
+  private MixedOperation<Crd, CrdList, CrdDoneable, Resource<Crd, CrdDoneable>> crdClient;
+
+  protected AbstractCrdController(KubernetesClient client, String crdPath, Long resyncCycle) {
+    this.client = client;
+    this.crdPath = crdPath;
+    this.resyncCycle = resyncCycle;
+
+    blockingQueue = new ArrayBlockingQueue<>(WORKING_QUEUE_SIZE);
+
+    if (this.client == null) {
+      this.client = new DefaultKubernetesClient();
+    }
+
+    namespace = this.client.getNamespace();
+
+    if (namespace == null) {
+      namespace = "default";
+      LOGGER.trace("No namespace found via config, assuming " + namespace);
```

Comment:
Also only going to comment once: When you use SLF4J this can be changed to

```suggestion
      LOGGER.trace("No namespace found via config, assuming {}", namespace);
```